### PR TITLE
Changes to number arguments

### PR DESCRIPTION
**Pull request by bartekn**
Changes this PR introduces:
### Change in amounts intepretation (7 fixed decimal places)
Following code:
```js
var destination = "GCEZWKCA5VLDNRLN3RPRJMRZOX3Z6G5CHCGSNFHEYVXM3XOJMDS674JZ";
var amount = "1000";
var asset = new StellarBase.Asset("USD", "GDGU5OAPHNPU5UCLE5RDJHG7PXZFQYWKCFOEXSXNMR6KRQRI5T6XXCD7");
var operation = StellarBase.Operation.payment({destination, asset, amount});
```
will create XDR object with `amount` equal to `10000000000` (1000 * 10^7).
See also: https://github.com/stellar/stellar-protocol/issues/20

### Allow string, numbers and BigNumber objects in `price` arguments
All functions below will return the same XDR object.
```js
var opts = {};
opts.amount = '10';
opts.selling = new StellarBase.Asset("USD", "GDGU5OAPHNPU5UCLE5RDJHG7PXZFQYWKCFOEXSXNMR6KRQRI5T6XXCD7");
opts.buying = new StellarBase.Asset.native();
function fromString() {
   opts.price = '1.25'; // string
   return StellarBase.Operation.manageOffer(opts);
}
function fromNumber() {
   opts.price = 1.25; // number
   return StellarBase.Operation.manageOffer(opts);
}
function fromBigNumber() {
   opts.price = new BigNumber(5).dividedBy(4);
   return StellarBase.Operation.manageOffer(opts);
}
```
This change closes https://github.com/stellar/js-stellar-base/issues/26.

### Add tests to `best_r` function

Tests thanks to BigNumber's [toFraction](https://github.com/MikeMcl/bignumber.js/blob/master/test/toFraction.js)

### CODE REVIEW

File: src/operation.js [hunk 1]
```diff
@@ -154,8 +162,8 @@ export class Operation {
         if (!isUndefined(opts.limit) && !isString(opts.limit)) {
             throw new TypeError('limit argument must be of type String');
         }
-        let limit           = opts.limit ? opts.limit : "9223372036854775807";
-        attributes.limit    = Hyper.fromString(limit);
+        let limit           = opts.limit ? new BigNumber(opts.limit).mul(ONE) : new BigNumber(MAX_INT64);
```
Comment (bartekn) [inline src/operation.js]: That's the only thing I'm not sure of. Are we using the same convention (https://github.com/stellar/stellar-protocol/issues/20) here too?

**Comment by nullstyle** [comment]
The replacement of `best_r` is not acceptable in its present form.  We must ensure that _both_ the denominator and the numerator are valid.

File: src/operation.js [hunk 2]
```diff
@@ -34,9 +38,10 @@ export class Operation {
             throw new TypeError('startingBalance argument must not be empty');
         }
         let attributes = {};
-        attributes.destination  = Keypair.fromAddress(opts.destination).accountId();
-        attributes.startingBalance = Hyper.fromString(opts.startingBalance);
-        let createAccount = new xdr.CreateAccountOp(attributes);
+        attributes.destination     = Keypair.fromAddress(opts.destination).accountId();
+        let startingBalance        = new BigNumber(opts.startingBalance).mul(ONE);
```
Comment (nullstyle) [inline src/operation.js]: I recommend that you extract the right hand side of this statement into a helper function.  In ruby, I named it  `interpret_amount`.

File: src/operation.js [hunk 3]
```diff
@@ -72,8 +77,9 @@ export class Operation {
         let attributes = {};
         attributes.destination  = Keypair.fromAddress(opts.destination).accountId();
         attributes.asset        = opts.asset.toXdrObject();
-        attributes.amount       = Hyper.fromString(opts.amount);
-        let payment = new xdr.PaymentOp(attributes);
+        let amount              = new BigNumber(opts.amount).mul(ONE);
```
Comment (nullstyle) [inline src/operation.js]: You should use your `interpret_amount` helper method here.

File: src/operation.js [hunk 4]
```diff
@@ -34,9 +38,10 @@ export class Operation {
             throw new TypeError('startingBalance argument must not be empty');
         }
         let attributes = {};
-        attributes.destination  = Keypair.fromAddress(opts.destination).accountId();
-        attributes.startingBalance = Hyper.fromString(opts.startingBalance);
-        let createAccount = new xdr.CreateAccountOp(attributes);
+        attributes.destination     = Keypair.fromAddress(opts.destination).accountId();
+        let startingBalance        = new BigNumber(opts.startingBalance).mul(ONE);
+        attributes.startingBalance = Hyper.fromString(startingBalance.toString());
```
Comment (nullstyle) [inline src/operation.js]: Ideally, this line would look like the following:

`attributes.startingBalance = this.interpretAmount(opts.startingBalance);`

File: src/operation.js [hunk 5]
```diff
@@ -122,12 +128,14 @@ export class Operation {
 
         let attributes = {};
         attributes.sendAsset    = opts.sendAsset.toXdrObject();
-        attributes.sendMax      = Hyper.fromString(opts.sendMax);
+        let sendMax             = new BigNumber(opts.sendMax).mul(ONE);
+        attributes.sendMax      = Hyper.fromString(sendMax.toString());
```
Comment (nullstyle) [inline src/operation.js]: You should use your `interpret_amount` helper method here.

File: src/operation.js [hunk 6]
```diff
@@ -122,12 +128,14 @@ export class Operation {
...
         attributes.destination  = Keypair.fromAddress(opts.destination).accountId();
         attributes.destAsset    = opts.destAsset.toXdrObject();
-        attributes.destAmount   = Hyper.fromString(opts.destAmount);
+        let destAmount          = new BigNumber(opts.destAmount).mul(ONE);
```
Comment (nullstyle) [inline src/operation.js]: You should use your `interpret_amount` helper method here.

File: src/operation.js (hunk 1 above)
Comment (nullstyle) [inline src/operation.js]: yep!

File: src/operation.js [hunk 7]
```diff
@@ -310,14 +318,16 @@ export class Operation {
         if (!isString(opts.amount)) {
             throw new TypeError('amount argument must be of type String');
         }
-        attributes.amount = Hyper.fromString(opts.amount);
+        let amount = new BigNumber(opts.amount).mul(ONE);
```
Comment (nullstyle) [inline src/operation.js]: You should use your `interpret_amount` helper method here.

File: src/operation.js [hunk 8]
```diff
@@ -358,14 +368,16 @@ export class Operation {
         if (!isString(opts.amount)) {
             throw new TypeError('amount argument must be of type String');
         }
-        attributes.amount = Hyper.fromString(String(opts.amount));
+        let amount = new BigNumber(opts.amount).mul(ONE);
```
Comment (nullstyle) [inline src/operation.js]: You should use your `interpret_amount` helper method here.

File: src/operation.js [hunk 9]
```diff
@@ -358,14 +368,16 @@ export class Operation {
...
             throw new TypeError('price argument is required');
         }
-        let approx = best_r(opts.price);
+        let price = new BigNumber(opts.price);
+        let approx = price.toFraction(MAX_DENOMINATOR);
         attributes.price = new xdr.Price({
-            n: approx[0],
-            d: approx[1]
+            n: parseInt(approx[0]),
+            d: parseInt(approx[1])
```
Comment (nullstyle) [inline src/operation.js]: This block above is copy-pasta from the line 320-ish area.  Please extract into a helper method.

**Comment by nullstyle** [comment]
Things are looking good, but as reflected in the comments I'm concerned with the amount of copy-pasta in this PR.

**Comment by bartekn** [comment]
Reverted to `best_r` and added some tests. Should be better now.

File: src/operation.js [hunk 10]
```diff
@@ -34,16 +38,16 @@ export class Operation {
             throw new TypeError('startingBalance argument must not be empty');
         }
         let attributes = {};
-        attributes.destination  = Keypair.fromAddress(opts.destination).accountId();
-        attributes.startingBalance = Hyper.fromString(opts.startingBalance);
-        let createAccount = new xdr.CreateAccountOp(attributes);
+        attributes.destination     = Keypair.fromAddress(opts.destination).accountId();
+        let startingBalance        = this._toStellarAmount(opts.startingBalance);
+        attributes.startingBalance = Hyper.fromString(startingBalance.toString());
```
Comment (nullstyle) [inline src/operation.js]: I think perhaps I didn't comment well in the last round.  Notice how everytime there is a call to `this._toStellarAmount` we immediately take the result and pass it into `Hyper.fromString`?  That's a smell that you didn't extract enough of the commonalities into the function.

The 2 lines above this comment should be a single line that reads: `attributes.startingBalance = this._toStellarAmount(opts.startingBalance)`

`_toStellarAmount` should return a Hyper.

File: src/operation.js [hunk 11]
```diff
@@ -438,31 +430,33 @@ export class Operation {
         }
 
         let attrs = operation.body().value();
+        let n, price;
         switch (operation.body().switch().name) {
             case "createAccount":
                 result.type = "createAccount";
                 result.destination = accountIdtoAddress(attrs.destination());
-                result.startingBalance = attrs.startingBalance().toString();
+                result.startingBalance = new BigNumber(attrs.startingBalance()).div(ONE).toString();
```
Comment (nullstyle) [inline src/operation.js]: Whole 'nother round of copy-pasta going on here.   Extract to a function called `_fromStellarAmount`.

File: src/operation.js [hunk 12]
```diff
@@ -438,31 +430,33 @@ export class Operation {
...
                 break;
             case "payment":
                 result.type = "payment";
                 result.destination = accountIdtoAddress(attrs.destination());
                 result.asset = Asset.fromOperation(attrs.asset());
-                result.amount = attrs.amount().toString();
+                result.amount = new BigNumber(attrs.amount()).div(ONE).toString();
```
Comment (nullstyle) [inline src/operation.js]: `_fromStellarAmount`

File: src/operation.js [hunk 13]
```diff
@@ -438,31 +430,33 @@ export class Operation {
...
                 break;
             case "pathPayment":
                 result.type = "pathPayment";
                 result.sendAsset = Asset.fromOperation(attrs.sendAsset());
-                result.sendMax = attrs.sendMax().toString();
+                result.sendMax = new BigNumber(attrs.sendMax()).div(ONE).toString();
```
Comment (nullstyle) [inline src/operation.js]: `_fromStellarAmount`

File: src/operation.js [hunk 14]
```diff
@@ -438,31 +430,33 @@ export class Operation {
...
                 result.destination = accountIdtoAddress(attrs.destination());
                 result.destAsset = Asset.fromOperation(attrs.destAsset());
-                result.destAmount = attrs.destAmount().toString();
+                result.destAmount = new BigNumber(attrs.destAmount()).div(ONE).toString();
```
Comment (nullstyle) [inline src/operation.js]: `_fromStellarAmount`

File: src/operation.js [hunk 15]
```diff
@@ -438,31 +430,33 @@ export class Operation {
...
                 result.path = attrs.path();
                 break;
             case "changeTrust":
                 result.type = "changeTrust";
                 result.line = Asset.fromOperation(attrs.line());
-                result.limit = attrs.limit().toString();
+                let limit = new BigNumber(attrs.limit().toString()).div(ONE);
+                result.limit = limit.toString();
```
Comment (nullstyle) [inline src/operation.js]: `_fromStellarAmount`

File: src/operation.js [hunk 16]
```diff
@@ -496,17 +490,20 @@ export class Operation {
                 result.type = "manageOffer";
                 result.selling = Asset.fromOperation(attrs.selling());
                 result.buying = Asset.fromOperation(attrs.buying());
-                result.amount = attrs.amount().toString();
-                result.price = attrs.price().n() / attrs.price().d();
+                result.amount = new BigNumber(attrs.amount()).div(ONE).toString();
```
Comment (nullstyle) [inline src/operation.js]: `_fromStellarAmount`

File: src/operation.js [hunk 17]
```diff
@@ -496,17 +490,20 @@ export class Operation {
...
                 result.offerId = attrs.offerId().toString();
                 break;
             case "createPassiveOffer":
                 result.type = "createPassiveOffer";
                 result.selling = Asset.fromOperation(attrs.selling());
                 result.buying = Asset.fromOperation(attrs.buying());
-                result.amount = attrs.amount().toString();
-                result.price = attrs.price().n() / attrs.price().d();
-
+                result.amount = new BigNumber(attrs.amount()).div(ONE).toString();
```
Comment (nullstyle) [inline src/operation.js]: `_fromStellarAmount`

**Comment by nullstyle** [comment]
@bartekn I've finished another round of comments.  Almost there!

**Comment by bartekn** [comment]
Done.

**Comment by nullstyle** [comment]
@bartek :+1:

**Pull request merged**